soc/xtensa/intel_adsp/tools/cavstool.py

```diff
@@ -546,6 +548,45 @@ def winstream_read(last_seq):
             # Found to be useful when it really goes wrong
             return (seq, result.decode("utf-8", "replace"))
 
+def read_from_gdb():
+    rx_head = bar4_mmap[DEBUG_OFFSET + 256]
+    rx_tail = bar4_mmap[DEBUG_OFFSET + 256 + 64]
+    msg = ""
+    while rx_head != rx_tail:
+        msg += chr(bar4_mmap[DEBUG_OFFSET + 256 + 128 + rx_tail])
+        rx_tail = (rx_tail + 1) % RING_SIZE
+    bar4_mmap[DEBUG_OFFSET + 256 + 64] = rx_tail
+    if msg:
+        os.write(gdb_client_port, msg.encode("ascii"))
+
+
+RING_SIZE = 128
+def write_to_gdb():
+    msg = os.read(gdb_client_port, 128).decode()
+    if len(msg) > 0:
+        # Write message to SRAM ringbuffer
+        head = bar4_mmap[DEBUG_OFFSET]
+        tail = bar4_mmap[DEBUG_OFFSET + 64]
+        tx_data_offset = DEBUG_OFFSET + 128
+        for c in msg:
+            while (head + 1) % RING_SIZE == tail:
+                bar4_mmap[DEBUG_OFFSET] = head
+                tail = bar4_mmap[DEBUG_OFFSET + 64]
+
+            bar4_mmap[tx_data_offset + head] = ord(c)
+            head = (head + 1) % RING_SIZE
+
+        bar4_mmap[DEBUG_OFFSET] = head
+
+
+def create_gdb_pty():
+    global gdb_client_port
+    # The user connects their gdb client to user_port, and this program accesses
+    # that data through gdb_client_port
+    (gdb_client_port, user_port) = pty.openpty()
+    name = os.ttyname(user_port)
+    log.info(f"GDB PTY at: {name}")
+    asyncio.get_event_loop().add_reader(gdb_client_port, write_to_gdb)
```

Comment:
Wouldn't it make more sense (and be more portable) to open a TCP socket for gdb instead?  Cavstool (which is getting WAY too big) has already grown support for server sockets.

Reply:
> Cavstool (which is getting WAY too big) has already grown support for server sockets.

Yes, and it's been a very unreliable mess:
https://github.com/zephyrproject-rtos/zephyr/issues/46864#issuecomment-1167545791

Just for completeness, I have NOT looked at this PR. This particular comment just caught my attention.